Metrics: quick start docs for Prometheus/Grafana (#2795)
open-telemetry/opentelemetry-dotnet: docs/metrics/learning-prometheus-http-server/README.md

- [Quick start on exporting metrics to Prometheus/Grafana](#quick-start-on-exporting-metrics-to-prometheusgrafana)
  - [Prerequisite](#prerequisite)
  - [Introduction](#introduction)
  - [Get Prometheus](#get-prometheus)
  - [Prometheus HTTP server](#prometheus-http-server)
    - [Configuration](#configuration)
    - [Start Prometheus](#start-prometheus)
    - [Configure OpenTelemetry to Expose metrics to Prometheus Endpoint](#configure-opentelemetry-to-expose-metrics-to-prometheus-endpoint)
    - [Check Results in Prometheus](#check-results-in-prometheus)
    - [View/Query Results with Grafana](#viewquery-results-with-grafana)

## Prerequisite

It is highly recommended to go over the [getting-started](../getting-started/)
project under the metrics document folder before following along this document.

## Introduction

- [What is Prometheus?](https://prometheus.io/docs/introduction/overview/)

- [Grafana support for
  Prometheus](https://prometheus.io/docs/visualization/grafana/#creating-a-prometheus-graph)

## Get Prometheus

Follow the [first steps]((https://prometheus.io/docs/introduction/first_steps/))
in the prometheus official document.

> ## Prometheus HTTP server

Reviewer: Curious - is "Prometheus HTTP server" a typical term folks would use?

Author: Not really. In the prometheus.io official doc, authors use Prometheus server there. I'll update remove the HTTP here.

### Configuration

After downloading the [latest release](https://prometheus.io/download/), extract
it to a local location that's easy to access. We will find the default

> prometheus configuration yaml file in the folder, named `prometheus.yml`.

Reviewer: ```suggestion
Prometheus configuration YAML file in the folder, named `prometheus.yml`.
```

> Replace all the content with:

Reviewer: Maybe instead of replace, create a new file? (so if people made newbie mistake, they won't turn down the default Prometheus config)

Author: Updated as suggested!

```
global:
  scrape_interval: 10s
  scrape_timeout: 10s
  evaluation_interval: 10s
scrape_configs:

> - job_name: OpenTelemetryTest # assign a job name, as prometheus server can listen to multiple jobs

Reviewer: ```suggestion
- job_name: OpenTelemetryTest
```

Reviewer: Or maybe something "MyOpenTelemetryDemo" (something that is self-explanatory)

Author: Updated.

  honor_timestamps: true
  scrape_interval: 1s
  scrape_timeout: 1s
  metrics_path: /metrics
  scheme: http
  follow_redirects: true
  static_configs:
  - targets:
    - localhost:9184 # set the port that prometheus server will listen to
```

### Start Prometheus

Follow the instructions from
[starting-prometheus](https://prometheus.io/docs/introduction/first_steps/#starting-prometheus)

> to start the prometheus server and verify it has been started successfully.

Reviewer: ```suggestion
to start the Prometheus server and verify it has been started successfully.
```

Once the server is started, we are going to make some small tweaks to the
example in the getting-started metrics [example](../getting-started/Program.cs)
to export our metrics to the endpoint that prometheus was configured to listen
to.

### Configure OpenTelemetry to Expose metrics to Prometheus Endpoint 

Create a new console application and run it:

```sh
dotnet new console --output prometheus-http-server
cd prometheus-http-server
dotnet run
```

We will have to add a reference to prometheus exporter to the
prometheus-http-server application.

```shell
dotnet add package OpenTelemetry.Exporter.Prometheus --version 1.2.0-rc1
```

Now, replace the below line in the [Program.cs](../getting-started/Program.cs)

```csharp
.AddConsoleExporter()
```

with

```csharp
.AddPrometheusExporter(opt =>
{
    opt.StartHttpListener = true;
    opt.HttpListenerPrefixes = new string[] { $"http://localhost:9184/" };
})
```

We set the options for our prometheus exporter to export data via the endpoint

> that we've configured prometheus server to listen to in the prometheus.yml file.

Reviewer: ```suggestion
that we've configured Prometheus server to listen to in the `prometheus.yml` file.
```

Also, for our learning purpose, use a while-loop to keep increasing the counter
value until any key is pressed.

```csharp
Console.WriteLine("Press any key to exit");
while (!Console.KeyAvailable)
{
    Thread.Sleep(1000);
    MyFruitCounter.Add(1, new("name", "apple"), new("color", "red"));
    MyFruitCounter.Add(2, new("name", "lemon"), new("color", "yellow"));
    MyFruitCounter.Add(1, new("name", "lemon"), new("color", "yellow"));
    ...
    ...
    ...
}
```

> After the above modifications, now our `program.cs` should look like [this](./Program.cs).

Reviewer: ```suggestion
After the above modifications, now our `Program.cs` should look like [this](./Program.cs).
```

### Check Results in Prometheus

> Execute the application and leave the process running in the background. Now we

Reviewer: ```suggestion
Start the application and leave the process running. Now we
```

I guess "in the background" might generate some confusion since it could mean different things depending on the operating system.
